dual_net.py

```diff
@@ -134,6 +134,40 @@ def should_log(i):
                     logger.add_summary(weight_summaries, global_step)
             self.save_weights()
 
+    def validate(self, tf_records, batch_size=128, logdir=None, init_from=None, num_steps=1000):
+        if logdir is None:
+            print("Need to run eval with a logdir")
+            return
+        if not logdir.endswith('_eval'):
+            logdir += "_eval"
+
+        with self.sess.graph.as_default():
+            input_tensors = preprocessing.get_input_tensors(
+                TRAIN_BATCH_SIZE, tf_records)
+            output_tensors = dual_net(input_tensors, TRAIN_BATCH_SIZE,
+                                      train_mode=False, **self.hparams)
+            train_tensors = train_ops(
+                input_tensors, output_tensors, **self.hparams)
+
+            # just run our cost tensors
+            validate_tensors = {k: train_tensors[k] for k in (
+                'policy_cost', 'value_cost', 'l2_cost',
+                'combined_cost')}
+            self.initialize_weights(init_from)
+            training_stats = StatisticsCollector()
+            logger = tf.summary.FileWriter(logdir, self.sess.graph)
+
+            for i in tqdm(range(num_steps)):
+                try:
+                    tensor_values = self.sess.run(validate_tensors)
+                except tf.errors.OutOfRangeError:
+                    break
+                training_stats.report(tensor_values)
+
+                if i % 10 == 9:
+                    accuracy_summaries = training_stats.collect()
+                    logger.add_summary(accuracy_summaries, i)
```

Comment:
the second arg of add_summary is global_step; you want to fetch the global step by doing sess.run(train_tensors['global_step']). That way, the eval metrics show up correctly aligned to the train metrics.

Reply:
yeah so my thought on this:  a run with ~hundreds of checkpoints can produce so much TB data that it takes ~30-45 minutes for tensorboard to chew through it all.  My thought was that by publishing validation stats for each new model with the same x-domain, they would overlay pretty easily and it would be easier to see overfitting.

I think having everything in an -eval directory and only having one point per model will work just as well, though, plus being able to align it with the training metrics is a big win.  Will do.

Reply:
interesting - my previous long runs have taken 1-2 minutes to load onto TB, but I guess these are even longer runs.

I would reduce the collect() frequency so that instead of having a gazillion noisy summary entries, we just have fewer summary entries averaged over longer time periods.

I don't think the overlaying is going to work the way you described - I think as is, everything will be overlaid where the X axis is 1-1000, and the 1000 y-values will literally be random noise centered around some value. Should collect() once at end and grab the global step and add_summary() once. That way the validation will plot nicely over time / over generations.

Reply:
agreed, bad idea, done.